### ISSUE TITLE
2020 캐글러닝데이 (12/17)

### ISSUE THREAD

**Issue by sallyy1** [sallyy1/Data-Analysis#9]
### **1. 클래스 불균형 문제 해결을 위한 다양한 방법론 소개(Class imbalanced problems & Feature Engineering)**
https://www.youtube.com/watch?v=HvOnWPnHISM&feature=youtu.be

*Preprocessing 단계 - EDA
1. Removal or interpolate missing values 
2. Raw data scaling
3. **Data Interpretation** (데이터에 대한 이해가 중요)

3-1) _변수 생성_
 - 사람에 의한 사전지식 반영
 - 데이터 의존 (Clustering 등)

3-2) _차원 축소_
 - 변수 선택 (Lasso, GA, SA 등)
 - 변수 압축 (PCA, AE 등)
 
3-3) _빈도수 조정_
 - Oversampler (**SMOTE** 등) : 학습비용이 증가
 -  Underersampler (Near miss 등) : 학습비용이 감소, 실제 존재하던 샘플을 제거하는 행위 -> 분포를 보장할 수 없다는 단점
(*Resampler는 좋은 성능을 보여주지만, 최선의 대안은 아님)

*모델 하이퍼파라미터 튜닝 < 변수 다양화 및 데이터 상 다양한 의미 찾기
1) 인과관계에 따른 설명 가능한 변수 생성
- Capital loss/ Capital gain 등
- 두 feature 간 교집합 => 새로운 feature 생성

2) 인과관계에 따른 설명 가능한 학습 셋 구성
- confusion matrix를 각 feature별로 표현해보기 (TN, FP, FN, TP) => 에러가 커지는 특정 feature에서 특정 값을 찾을 수 있음
- Capital loss / Capital gain 에 따른 학습 셋 별도로 구성해보기

3) 데이터 압축
- Denoisiong Autoencoder (Encoder - Latent Vector - Decoder)

--------------------------------------
### **2. CodeBERT를 이용하여 Code Diff이 주어졌을때 그 코드를 자연어로 요약하는 프로젝트(codediff2NL)**
https://www.youtube.com/watch?v=GOyrsR2aW-Y&feature=youtu.be

*commit-autosuggestions(CodeBERT를 기반으로 커밋 메세지를 생성해주는 모델)
https://github.com/graykode/commit-autosuggestions (깃헙 주소)

*PLM (Pretrained Language Model)
1) _Transformer(2017)_
 1. Positional Encoding
 2. Scaled Dot-Product Attention
 3. Multi-Head Attention
 4. Feed Forward Network

2) _BERT(2018)_
 1. Task #1: Masked LM
 2. Tast #2: Next Sentence Prediction(NSP) -문장이 이어지는지 끊기는지 Binary Classification

3) _RoBERTa(2019)_
 1. 토크나이저를 Wordpiece에서 BPE(Byte Pair Encoding)로 변경
 2. Task #2 Next Sentence Prediction(NSP) 제거
 3. 매우 큰 배치 사이즈 (BERT:256 배치 -> 8K 배치) => 좋은 성능

4) _ELECTRA(2019)_
 - Generator + Discriminator
 - 다른 Language 모델에 비해 수렴이 빨리되는 장점

5) _CodeBERT(2020)_
 - NL Generator(w벡터) + Code Generator(c벡터)
 - 모델은 RoBERTa와 동일
 - Code2NL Task 있음

-----------------------
### **3. Weights & Biases 를 활용한 머신러닝 실험 모니터링하기**
https://www.youtube.com/watch?v=tqD9ovQVOo8&feature=youtu.be

- **wandb** Tool
*장점
• 학습 시 사용된 하이퍼 파라미터들을 저장할 수 있음
• 학습했던 각 실험들을 탐색하고, 비교하고, 시각화할 수 있음
• 학습 환경의 시스템을 어떻게 사용하고 있는지 분석할 수 있음
• 다른 사람들과 협업할 수 있음
• 과거 실험 결과들을 복제할 수 있음
• 하이퍼 파라미터 튜닝이 가능함
• 실험했던 모든 기록들을 영구적으로 저장할 수 있음

*wandb tutorial
• https://colab.research.google.com/github/wandb/examples/blob/master/colabs/intro/Intr
o_to_Weights_%26_Biases.ipynb
• https://pytorch.org/tutorials/beginner/dcgan_faces_tutorial.html

*Reference
• wandb slack channel
• wandb github page: https://github.com/wandb/client
• wandb tutorial:
https://colab.research.google.com/github/wandb/examples/blob/master/colabs/intro/Intro_to_Weights_%26_Biases.ipynb
• kaggle report: https://storage.googleapis.com/kagglemedia/surveys/Kaggle%20State%20of%20Machine%20Learning%20and%20Data%20Science%202020.pdf
• code block: https://carbon.now.sh/
• pytorch DCGAN tutorial: https://pytorch.org/tutorials/beginner/dcgan_faces_tutorial.html

--------------------
### **4. 파이썬 몇 줄로 간단히 해보는 탐색적 자료 분석**
https://www.youtube.com/watch?v=jXXyGAYXj1k&feature=youtu.be

*Tidy data 용어

*EDA 시각화 라이브러리 소개
(pip install 가능)

1) sweetviz
https://pypi.org/project/sweetviz/
-sv.analyze(df_train)함수

2) pandas-profiling
https://pypi.org/project/pandas-profiling/
-ProfileReport(df_train, title='Pandas Profiling Report', explorative=True)

3) pandasgui (마우스 클릭)
https://pypi.org/project/pandasgui/
